src/GraphPanel.java

```diff
@@ -184,30 +173,42 @@ private void drawGrid(Graphics2D g2d, int width, int height, double step) {
         }
     }
 
-    private void drawFunction(Graphics2D g2d, int width) {
-        GeneralPath path = new GeneralPath();
+    private void drawFunctions(Graphics2D g2d, int width) {
+        GeneralPath path;
 
         double minT = toWorldCoordinates(new Vector2D(0, 0)).x;
         double maxT = toWorldCoordinates(new Vector2D(width, 0)).x;
 
-        int numSteps = 500;
+        int numSteps = (int) Math.max(5000, 10000 * zoom);
```

Comment:
numSteps should not depend on the zoom level. For parametric functions, the user must provide the number of steps. For polynomials, the bounds for t are set dynamically by the left and right borders, anyway. The number of segments should only consider the screen width in this case. (x pixels/segment)